### PR TITLE
Add UI for NetworkPolicies

### PR DESCRIPTION
**Pull request by bashofmann**
NetworkPolicies are one of the more complex Kubernetes resources. But they are also frequently used in enterprise contexts to restrict incoming and outgoing traffic of pods. A UI would make this much more accessible, especially for Kubernetes beginners.

Some screenshots:

Bildschirmfoto 2021-12-08 um 11 36 54
Bildschirmfoto 2021-12-08 um 11 37 03
Bildschirmfoto 2021-12-08 um 11 37 14
Bildschirmfoto 2021-12-08 um 11 39 30
Bildschirmfoto 2021-12-08 um 11 39 49

Adresses: #4729

### CODE REVIEW

File: edit/networking.k8s.io.networkpolicy/PolicyRulePort.vue [hunk 1]
```diff
@@ -0,0 +1,50 @@
...
+<template>
+  <div class="row">
+    <div class="col span-4">
+      <LabeledInput
+        v-model.number="value.port"
+        :mode="mode"
+        type="number"
+        min="1"
+        max="65535"
+        placeholder="e.g. 8080"
```
Comment (richard-cox) [inline edit/networking.k8s.io.networkpolicy/PolicyRulePort.vue:36]: Should be moved to en-us

File: edit/networking.k8s.io.networkpolicy/PolicyRuleTarget.vue [hunk 2]
```diff
@@ -0,0 +1,281 @@
...
+    podSelectorExpressions:       'updateMatches',
+    namespaceSelectorExpressions: 'updateMatches',
+  },
+  methods: {
+    validateCIDR() {
+      const exceptCidrs = this.value[TARGET_OPTION_IP_BLOCK].except || [];
+
+      this.invalidCidrs = [];
+
+      const regex = /^((25[0-5]|2[0-4][0-9]|[01]?[0-9][0-9]?)\.){3}(25[0-5]|2[0-4][0-9]|[01]?[0-9][0-9]?)\/(3[0-2]|2[0-9]|1[0-9]|[0-9])$/;
```
Comment (richard-cox) [inline edit/networking.k8s.io.networkpolicy/PolicyRuleTarget.vue:149]: Although overkill for this use case, validating a cidr might be useful elsewhere (i think the harvester team have hit this). Could this be made into a `@/utils/validators/cidr.js` so it's in a common place ready to use next time?

File: edit/networking.k8s.io.networkpolicy/PolicyRule.vue [hunk 3]
```diff
@@ -0,0 +1,103 @@
...
+      type:    Array,
+      default: () => {
+        return [];
+      },
+    },
+  },
+  methods: {
+    onAddPort() {
+      if (this.$refs.lastPort?.focus) {
+        this.$refs.lastPort.focus();
```
Comment (richard-cox) [inline edit/networking.k8s.io.networkpolicy/PolicyRule.vue:47]: I didn't see this or the one below working. These are convenience methods though, so worst case they can be removed

**Review by richard-cox — CHANGES_REQUESTED** [review]
A pretty solid PR for the number of lines, nice!

There's a design question regarding the location of ingress/egress rules and a few niggles, otherwise pretty good

- I (locally) rebased the branch with `master`, it's worth doing this before merging to improve the git tree.
- The ingress/egress rules look rather orphaned below the main tabs content. @lvuch do we have a pattern for this kind of thing (tab lists in tabs?), would they be better appearing/disappearing as ingress & egress tabs that would themselves contain the tab lists? (see issue for screen shots or ping me for details)
- Given that empty/initial ingress/egress rules aren't rejected, clicking on either checkbox to show either control should pre-populate the collections with an initial rule (there may be a setting on the component that does this automatically)
- The pods list was always empty, even when the banner text listed matching pods. Probably a reactive issue (normally solved by using Vue.set/this.$set on the property that has changed)
- Bug - Create a policy with the ingress/egress options checked, save policy, edit policy, click `+` on ingress rule
  ```
  client.js?06a0:97 TypeError: Cannot read properties of undefined (reading 'push')
    at VueComponent.addPolicyRule (PolicyRules.vue?0a66:42)
  ```
  (does not happen if there are existing rules)

File: edit/networking.k8s.io.networkpolicy/PolicyRulePort.vue (hunk 1 above)
Comment (bashofmann) [reply edit/networking.k8s.io.networkpolicy/PolicyRulePort.vue]: Done

File: edit/networking.k8s.io.networkpolicy/PolicyRuleTarget.vue (hunk 2 above)
Comment (bashofmann) [reply edit/networking.k8s.io.networkpolicy/PolicyRuleTarget.vue]: Done

File: edit/networking.k8s.io.networkpolicy/PolicyRule.vue (hunk 3 above)
Comment (bashofmann) [reply edit/networking.k8s.io.networkpolicy/PolicyRule.vue]: I removed these

File: edit/networking.k8s.io.networkpolicy/index.vue [hunk 4]
```diff
@@ -0,0 +1,258 @@
...
+    PolicyRules,
+    ResourceTable,
+    Tab,
+    Tabbed,
+  },
+
+  mixins: [CreateEditView],
+
+  async fetch() {
+    this.allPods = await this.$store.dispatch('cluster/findAll', { type: POD });
```
[the remark this answers is not in the document]
Comment (bashofmann) [reply edit/networking.k8s.io.networkpolicy/index.vue]: Done

**Comment by bashofmann** [comment]
> I (locally) rebased the branch with master, it's worth doing this before merging to improve the git tree.

PR is rebased.

> The ingress/egress rules look rather orphaned below the main tabs content. @lvuch do we have a pattern for this kind of thing (tab lists in tabs?), would they be better appearing/disappearing as ingress & egress tabs that would themselves contain the tab lists? (see issue for screen shots or ping me for details)

Should I change anything here? How should it look?

> Given that empty/initial ingress/egress rules aren't rejected, clicking on either checkbox to show either control should pre-populate the collections with an initial rule (there may be a setting on the component that does this automatically)

An empty set of rules is actually fine and should work.

Without checking the ingress or egress box, all traffic is allowed. If you check a box and don't define any allowed rules, this would mean that all ingress or egress traffic is forbidden.

> The pods list was always empty, even when the banner text listed matching pods. Probably a reactive issue (normally solved by using Vue.set/this.$set on the property that has changed)

No idea where I'd need to add this. For me it always updates and shows pods.

> Bug - Create a policy with the ingress/egress options checked, save policy, edit policy, click + on ingress rule
client.js?06a0:97 TypeError: Cannot read properties of undefined (reading 'push')
  at VueComponent.addPolicyRule (PolicyRules.vue?0a66:42)
(does not happen if there are existing rules)

Fixed

**Review by richard-cox — APPROVED** [review]
@bashofmann I've moved ingress & egress into their own tabs. This helps avoid lost content at the bottom that's a bit divorced from the checkboxes. I think our side we need to provide a better UX for this concept, but for the 2.6.4 timeframe this is fine. If you're happy with this change I'm happy to merge

**Comment by bashofmann** [comment]
@richard-cox looks great.

**Pull request merged**